src/core/environment.js

```diff
@@ -537,10 +537,12 @@ p5.prototype.pixelDensity = function(val) {
       this._pixelsDirty = true;
     }
     returnValue = this;
+    this.resizeCanvas(this.width, this.height, true); // as a side effect, it will clear the canvas
   } else {
     returnValue = this._pixelDensity;
   }
-  this.resizeCanvas(this.width, this.height, true);
+  // not resizing the canvas if we only read the pixelDensity
+  // -> fixes issue #3175
```

Comment:
these two comment lines above can be removed

Reply:
Agreed. I did only include the manual tests for easier integration of the pull request, so the integrator has some context.